### PR TITLE
Add small files check in garbage collection

### PR DESCRIPTION
**Pull request by hangc0276**
### Motivation
When we use `TransactionalEntryLogCompactor` to compact the entry log files, it will generate a lot of small entry log files, and for those files, the file usage is usually greater than 90%, which can not be compacted unless the file usage decreased.

image

### Changes
We introduce the entry log file size check during compaction, and the checker is controlled by `gcEntryLogSizeRatio`. 
If the total entry log file size is less than `gcEntryLogSizeRatio * logSizeLimit`, the entry log file will be compacted even though the file usage is greater than 90%. This feature is disabled by default and the `gcEntryLogSizeRatio` default value is `0.0`

### CODE REVIEW

**Comment by zymap** [comment]
Open an email to discuss it?

**Comment by hangc0276** [comment]
> Open an email to discuss it?

@zymap Ok, I send an email.

**Comment by hangc0276** [comment]
ping @merlimat @eolivelli @dlg99 @zymap Please help take a look at this PR, thanks.

**Review by eolivelli — APPROVED** [review]
great

+1

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/GarbageCollectorThread.java [hunk 1]
```diff
@@ -548,14 +548,24 @@ private boolean removeIfLedgerNotExists(EntryLogMetadata meta) throws EntryLogMe
...
+                + "on garbage collection.", conf.getGcEntryLogSizeRatio(), gcEntryLogSizeRatio,
+                gcEntryLogSizeRatio * conf.getEntryLogSizeLimit());
+        }
 
         final int numBuckets = 10;
         int[] entryLogUsageBuckets = new int[numBuckets];
         int[] compactedBuckets = new int[numBuckets];
 
         ArrayList<LinkedList<Long>> compactableBuckets = new ArrayList<>(numBuckets);
+        ArrayList<LinkedList<Long>> smallFilesCompactableBuckets = new ArrayList<>(numBuckets);
```
Comment (dlg99) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/GarbageCollectorThread.java:565]: I think you can avoid addition of another type of buckets.
Have something like useTargetEntryLogSizeForGc config;

Replace
```java
calculateUsageIndex(numBuckets, meta.getUsage());
```

with something like
```java
double usage = meta.getUsage();

if (conf.getUseTargetEntryLogSizeForGc()) {
  usage = (double) meta.getRemainingSize() / Math.max(meta.getTotalSize(), conf.getEntryLogSizeLimit());
}
```
and the rest will happen naturally.

As an improvement, transactional compaction can fallback to a regular one if the estimated size of entry log after compaction is below some limit.

Comment (hangc0276) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/GarbageCollectorThread.java]: @dlg99 Good suggestion. I updated the code with `useTargetEntryLogSizeForGc` flag.  Please help take a look, thanks.

> As an improvement, transactional compaction can fallback to a regular one if the estimated size of entry log after compaction is below some limit.

For this one, I will open a new PR to handle it.

**Review by zymap — APPROVED** [review]

**Review by horizonzy — APPROVED** [review]
LGTM.

**Comment by hangc0276** [comment]
This PR is hard to be cherry-picked to branch-4.14, and I will push another PR for branch-4.14

**Comment by zymap** [comment]
Cherry-pick this change without the tests class. The test class is based on the direct IO improvement.

**Pull request merged**